lib/elixir_bench/github/client.ex

```diff
@@ -1,44 +1,20 @@
 defmodule ElixirBench.Github.Client do
-  alias ElixirBench.Github.Client
+  @moduledoc """
+  The Github client specification.
 
-  defstruct [:base_url, ssl_options: []]
+  This module defines an interface for writing Github clients. The
+  clients provide the api to get the raw configuration based on the
+  benchee.yml file of the target repository. Their implementation can follow
+  different protocol interfaces, like the `ElixirBench.Github.ClientHTTP`
+  module or a client can be defined with testing pourposes, to be used as mock,
+  like in `ElixirBench.Github.ClientInMemory`.
 
-  def get_plain(%Client{} = client, url, params \\ []) do
-    get(client, url(client, url, params), [{"accept", "text/plain"}], &(&1))
-  end
+  A module that implements the Github client behaviour must export:
 
-  def get_json(%Client{} = client, url, params \\ []) do
-    get(client, url(client, url, params), [{"accept", "application/json"}], &Antidote.decode/1)
-  end
+    - a `get_yaml/1` function that takes the url path of the project and returns
+    a map with the configuration found in the benchee.yml that is stored
+    inside the `benchee directory`.
+  """
 
-  def get_yaml(%Client{} = client, url, params \\ []) do
-    get(client, url(client, url, params), [{"accept", "application/yaml"}], &decode_yaml/1)
-  end
-
-  @doc false
-  # Public for testing purposes
-  def decode_yaml(content) do
-    case :yamerl.decode(content, [:str_node_as_binary, map_node_format: :map]) do
-      [parsed] -> {:ok, parsed}
-      _ -> {:error, :invalid}
-    end
-  end
-
-  defp get(client, url, headers, cb) do
-    ssl_options = client.ssl_options
-    options = [:with_body, ssl_options: ssl_options]
-
-    case :hackney.get(url, headers, <<>>, options) do
-      {:ok, 200, _headers, data} ->
-        cb.(data)
-      {:ok, status, _headers, data} ->
-        {:error, {status, data}}
-      {:error, error} ->
-        {:error, error}
-    end
-  end
-
-  defp url(%{base_url: base}, url, params) do
-    Path.join(base, url) <> "?" <> URI.encode_query(params)
-  end
+  @callback get_yaml(path :: String) :: {:ok, config :: Map} | {:error, reason :: term}
```

Comment:
I don't believe these typespecs actually work but we don't really know whithout dialyzer setup :) Needs to be `String.t` and `Map.t` imo

Reply:
I will read more about how dialyzer and typespecs works. :man_student: